tests/test_tgis_backend.py

```diff
@@ -220,44 +251,107 @@ def test_local_tgis_unload(mock_tgis_proc_ok, tgis_mock_insecure):
+    # increase the health poll delay to greater than the load timeout
+    mock_tgis_server.health_delay = 1
     tgis_be = TGISBackend(
         {
             "local": {
-                "grpc_port": int(
-                    tgis_mock_insecure_health_delay.hostname.split(":")[-1]
-                ),
-                "http_port": tgis_mock_insecure_health_delay.http_port,
+                "grpc_port": int(mock_tgis_server.hostname.split(":")[-1]),
+                "http_port": mock_tgis_server.http_port,
                 "health_poll_delay": 0.01,
                 "health_poll_timeout": 0.01,
                 "load_timeout": 0.05,
             },
         }
     )
     assert tgis_be.local_tgis
-    assert not mock_tgis_proc_ok.called
+    assert not mock_tgis_fixture.server_launched()
 
-    # (For coverage!) make sure the health probe doesn't actually run
-    assert not tgis_be._tgis_proc_health_check()
+    # TODO: health check for coverage?
+    # # (For coverage!) make sure the health probe doesn't actually run
+    # assert not tgis_be._tgis_health_check()
 
     # Get a client handle and make sure that the server has launched
     with pytest.raises(RuntimeError):
         tgis_be.get_client("")
-    assert mock_tgis_proc_ok.called
+    assert mock_tgis_fixture.server_launched()
     assert tgis_be.local_tgis
     assert not tgis_be.model_loaded
 
 
+def test_local_tgis_autorecovery(mock_tgis_fixture: MockTGISFixture):
+    """Test that the backend can automatically restart the TGIS subprocess if it
+    crashes
+    """
+    # mock the subprocess to be our mock server and to come up working
+    mock_tgis_server: TGISMock = mock_tgis_fixture.mock_tgis_server
+    tgis_be = TGISBackend(
+        {
+            "local": {
+                "grpc_port": int(mock_tgis_server.hostname.split(":")[-1]),
+                "http_port": mock_tgis_server.http_port,
+                "health_poll_delay": 0.01,
+                "health_poll_timeout": 0.01,
+            },
+        }
+    )
+    assert tgis_be.local_tgis
+
+    # Get a client handle and make sure that the server has launched
+    tgis_client = tgis_be.get_client("")
+
+    assert mock_tgis_fixture.server_launched()
+
+    # requests should succeed
+    tgis_client.Generate(
+        generation_pb2.BatchedGenerationRequest(
+            requests=[
+                generation_pb2.GenerationRequest(text="Hello world"),
+            ],
+        ),
+    )
+
+    # "kill" the mock server
+    mock_tgis_server.stop()
+
+    # request should fail, which triggers the auto-recovery
+    with pytest.raises(grpc.RpcError):
+        tgis_client.Generate(
+            generation_pb2.BatchedGenerationRequest(
+                requests=[
+                    generation_pb2.GenerationRequest(text="Hello world"),
+                ],
+            ),
+        )
+
+    # wait for the server to reboot
+    # pause this thread to allow the reboot thread to start
+    time.sleep(0.5)
```

Comment:
Curious why we need this sleep here?

Reply:
The restart of TGIS is triggered by the failed request which launches a background thread. If `wait_until_ready()` is called before the background thread detects that TGIS is down, it will just return READY. There's probably a better way, but the `sleep` is meant to force a context switch so that the background thread gets a chance to detect the failure.

Reply:
got it, makes sense!